### PR TITLE
[inductor][NVGEMM] Define shared epilogue reduction contracts

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #194652
* #194651
* #194650
* #194649
* #194648
* #195539
* #195538
* #192928
* #190825
* #191515
* #191514
* #190822
* #192298
* #194873
* __->__ #195231
* #190821
* #194793
* #194792

<human note needed>

> Define a shared representation for grouped GEMM reductions and use it as the
> boundary between epilogue analysis and backend-specific code generation.
>
> ## Shared reduction representation
>
> `GemmReductionGeometry` describes which logical GEMM output axis is reduced
> and how many adjacent elements belong to each group. It also determines
> whether the reduction fits inside one TensorSSA fragment or requires physical
> cross-thread combination.
>
> `GemmReductionPlan` carries the complete semantic contract:
>
> - the grouped reduction geometry;
> - the associative primitive, such as sum, product, minimum, or maximum;
> - generated source, combine, finalize, and consumer callbacks;
> - the primary and auxiliary output buffers;
> - whether the reduction feeds the primary GEMM output; and
> - whether the generated tensor epilogue returns a fragment-local reduction.
>
> This separates the mathematical operation from the physical reduction
> mechanism. Backends implement fragment traversal, synchronization, and
> storage, while generated callbacks describe the computation applied at each
> stage.
>
> ## Loop IR analysis
>
> Extend the Loop IR analysis to recognize both explicit reduction nodes and
> reductions expressed as unrolled pointwise operations. For an unrolled
> reduction, the analysis:
>
> 1. Finds the repeated loads from the GEMM output.
> 2. Uses their index ranges and strides to infer the group size and M/N axis.
> 3. Verifies that every lane applies the same pointwise source expression.
> 4. Identifies the associative primitive.
> 5. Preserves the source expression for later CuTeDSL callback generation.
>
> For example, an expression equivalent to:
>
> ```python
> grouped = output.float().view(m, n // 4, 4)
> reduced = torch.exp(grouped).sum(dim=-1)
> ```
>
> is represented as an N-axis group-of-four sum whose source expression is
> `exp(value)`.
>
> ## FlexGEMM integration
>
> Replace FlexGEMM's parallel template-time reduction description with a wrapper
> around `GemmReductionPlan`. The FlexGEMM wrapper retains only provider-specific
> information: the auxiliary-output index, output layout, and operand-swap
> state.
>
> FlexGEMM lowering now passes `epilogue_analysis.reduction_plan` directly into
> that wrapper. The shared geometry also owns the fragment-width and
> physical-callback decision previously implemented by a FlexGEMM-specific
> helper.
>
> This gives FlexGEMM and NVGEMM the same semantic reduction plan while allowing
> each provider to retain its own runtime ABI and code-generation machinery.
>
> ## Relationship between FlexGEMM and NVGEMM
>
> Both backends use the same semantic reduction plan, while retaining their
> existing analysis and runtime interfaces:
>
> ```mermaid
> flowchart LR
>     FX["FlexGEMM FX epilogue"] --> FXA["GemmLocalReduceAnalysis"]
>     LOOP["NVGEMM scheduler Loop IR"] --> LIA["GemmEpilogueIRAnalysis"]
>
>     FXA --> PLAN["GemmReductionPlan<br/>geometry, primitive, callbacks, outputs"]
>     LIA --> CFG["GemmReductionConfig"]
>     CFG --> PLAN
>
>     PLAN --> FLEXCFG["FlexGemmEpilogueLocalReduceConfig<br/>plan + output index + layout + swap_ab"]
>     FLEXCFG --> FLEXRT["FlexGemmRuntimeLocalReducePlan"]
>     FLEXRT --> QUACK["FlexGEMM / QuACK"]
>
>     PLAN --> NVCAP["NVGEMM capability validation"]
>     NVCAP --> NVARGS["GemmReductionArguments"]
>     NVARGS --> NVKERNEL["Vendored NVGEMM kernel"]
> ```
>
> `GemmReductionPlan` is the shared semantic boundary. FlexGEMM adds only the
> output placement and layout information needed by its template, while NVGEMM
> binds the plan to runtime tensors and cache-specialization fields.
>
> ## Testing
>
> Unit coverage for the shared reduction plan and FlexGEMM adapter is included
> in this PR.
>
> End-to-end dense and block-scaled generated-epilogue coverage is provided by
> [#190825](https://github.com/pytorch/pytorch/pull/190825).
>
> ```bash
> lintrunner -a
> ```
>
> Description drafted with an AI assistant.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by drisspg — APPROVED** [review]

**Comment by mlazos** [comment]
## AI-Review

> *Automated multi-agent review (2 Claude + 2 Codex + GPT-5.6 verification). This is AI-generated — please verify findings manually.*

---

## Verified Issues
- **[medium] Config-to-plan conversion drops the selected reduction algorithm** — Location: `torch/_inductor/kernel/gemm_epilogue.py`, `GemmReductionPlan.from_config()`. Trigger: Call `from_config()` with a `GemmReductionConfig` whose `reduction_algorithm` is not `"default"`, without overriding it in `plan_fields`. Failure sequence: The factory copies other semantic fields but omits `config.reduction_algorithm`; the plan dataclass therefore applies its `"default"` value. Impact: The returned plan observably specifies a different physical reduction algorithm than the input configuration. Reviewers: Claude Design, Codex Design, Claude Bug, Codex Bug.
- **[medium] Synthetic matching accepts cross-lane expressions as per-element reduction sources** — Location: `torch/_inductor/kernel/loop_ir_epilogue_lowering.py`, `_expression_pattern()` and `_synthetic_reduction_element_ir()`. Trigger: A group-2 expression such as `(x[i] - x[j]) * (x[j] - x[i])`. Failure sequence: Source-load indices are erased, making the two subtraction terms appear identical; the matcher consequently returns a `"prod"` reduction with `x[i] - x[j]` as its purported per-element source. Impact: The helper returns a false reduction classification for an expression whose terms each depend on multiple lanes, violating its per-element-source contract and enabling incorrect lowering. Reviewers: Claude Design, Codex Bug.
